### PR TITLE
Add framerate limiting

### LINKED ISSUE #935: [feature] framerate cap settings (from crispy)
first mentioned in doomworld's thread

these are the PRs that prompted me to request the feature
https://github.com/fabiangreffrath/crispy-doom/pull/970
https://github.com/fabiangreffrath/crispy-doom/pull/988

### PR DESCRIPTION
**Pull request by mikeday0**
Closes #935.

### CODE REVIEW

File: src/m_misc.c [hunk 1]
```diff
@@ -2424,6 +2424,14 @@ default_t defaults[] = {
     "1 to enable uncapped rendering frame rate"
   },
 
+  // framerate limit
+  {
+    "fpslimit",
+    (config_t *) &fpslimit, NULL,
+    {0}, {35, 500}, number, ss_gen, wad_no,
```
Comment (fabiangreffrath) [inline src/m_misc.c:2431]: I'm afraid this won't work. The default value has to be in the min-max range.

Comment (mikeday0) [reply src/m_misc.c:2431]: Drat! How about I switch `uncapped` to a multivalue where 0 = off, 1 = uncapped with no limit, 2 = uncapped with limit?

Comment (fabiangreffrath) [reply src/m_misc.c:2431]: In the latter case, we could even uncap the upper limit. We should also mention in the description that this requires uncapped framerate enabled.

Comment (mikeday0) [reply src/m_misc.c:2431]: Sorry, I should have elaborated more. I definitely intend to expose this setting in the menu. There would be a separate line item to set the value for `fpslimit` and the `uncapped` menu item would grow from two options (OFF and ON) to three (OFF, ON no cap, ON with cap). If `uncapped` is set to OFF or ON no cap then I would gray out the `fpslimit` menu item.

The advantage to this is that we can cleanly specify the min-max range for `fpslimit` (no shenanigans with using a value of 0 to disable) and it does not disturb existing configs.

Comment (fabiangreffrath) [reply src/m_misc.c:2431]: Yes, this sounds like the best solution. The menu will need some rearrangement then (yay, it's high time, we didn't have this for literally *weeks* 😉).

Comment (mikeday0) [reply src/m_misc.c:2431]: I haven't dug into the Woof! menu code yet, but maybe we could do something without adding a new menu item. Could we perhaps have a press of the enter key bring up a prompt to allow the user to key in the frame limit value?

Comment (fabiangreffrath) [reply src/m_misc.c:2431]: Yes, of course. The Woof! menu has already inherited numerical input from Boom/MBF. For example, you may set a numerical value for game speed in percent.

File: src/m_menu.c [hunk 2]
```diff
@@ -3916,6 +3917,10 @@ static void M_SetMidiPlayer(void)
   S_RestartMusic();
 }
 
+static const char *uncapped_strings[] = {
```
Comment (fabiangreffrath) [inline src/m_menu.c:3920]: Hm, how do we handle this in Crispy? I mean, if the min/max range for the `fpslimit` value is the limiting factor here, we could just call a sanitizer function after that value has been set, which allows either `0` for "disable" or any other value higher than `35`.

Comment (mikeday0) [reply src/m_menu.c]: We coerce values from to 1 to 34 to 35 in Crispy. However for Woof I just implemented it so 1 through 34 get coerced to 0 instead. I think I might like this better. Any opinions? I may go back and update Crispy if the latter approach is preferred.

File: src/m_menu.c [hunk 3]
```diff
@@ -3916,6 +3917,24 @@ static void M_SetMidiPlayer(void)
...
+  }
+  else
+  {
+    gen_settings1[gen1_fpslimit].m_flags &= ~S_DISABLE;
+  }
+}
+
+static void M_CoerceFPSLimit(void)
+{
+  if (fpslimit > 0 && fpslimit < FPS_LIMIT_MIN)
```
Comment (fabiangreffrath) [inline src/m_menu.c:3934]: Why check for `> 0`? What if the value is e.g. `-1`? Granted, that's not possible due to the config system, but this restriction is strictly not necessary.

File: src/m_menu.c [hunk 4]
```diff
@@ -3932,7 +3951,10 @@ setup_menu_t gen_settings1[] = { // General Settings screen1
 
   // [FG] uncapped frame rate
   {"Uncapped Frame Rate", S_YESNO, m_null, M_X, M_Y+ gen1_uncapped*M_SPC,
-   {"uncapped"}},
+   {"uncapped"}, 0, M_EnableDisableFPSLimit},
+
+  {"FPS Limit", S_NUM, m_null, M_X,
```
Comment (fabiangreffrath) [inline src/m_menu.c:3956]: Should we call this "Frame Rate Limit" for consistency with the line above?

File: src/m_misc.c [hunk 5]
```diff
@@ -2424,6 +2424,14 @@ default_t defaults[] = {
     "1 to enable uncapped rendering frame rate"
   },
 
+  // framerate limit
+  {
+    "fpslimit",
+    (config_t *) &fpslimit, NULL,
+    {0}, {0, 500}, number, ss_gen, wad_no,
+    "framerate limit in frames per second (< 35 = Disable)"
```
Comment (fabiangreffrath) [inline src/m_misc.c:2432]: Why capital 'D' in "Disable"?

File: src/i_video.h [hunk 6]
```diff
@@ -80,6 +80,9 @@ extern int hires;      // killough 11/98
 
 extern int useaspect;
 extern int uncapped; // [FG] uncapped rendering frame rate
+
+#define FPS_LIMIT_MIN 35
```
Comment (fabiangreffrath) [inline src/i_video.h:84]: That's identical to `TICRATE` so let's just use that instead.

**Review by fabiangreffrath — CHANGES_REQUESTED** [review]
Works fine, thank you!

File: src/m_menu.c (hunk 3 above)
Comment (mikeday0) [reply src/m_menu.c:3934]: Yep, you are absolutely correct. This is vestigial code from when I copied over some of the Crispy implementation.

**Comment by liPillON** [comment]
Sorry for being pedantic, but after testing this I started wondering if the "uncapped framerate" option should be renamed to "framerate interpolation" to avoid ambiguity with the newly added "framerate limit" option.
What do you guys think?

**Comment by fabiangreffrath** [comment]
While "interpolation" is technically correct, it completely ignores the fact that this feature unchains Vanilla's 35 fps limit in the first place. We could call it "allow more than 35fps" or similar, but I think in the context of Doom source ports the term "uncapped framerate" is already pretty unambiguous.

**Pull request merged**